### PR TITLE
2.0 playlists code review

### PR DESCRIPTION
**Pull request by roxanariza**
User playlists v2

### CODE REVIEW

File: src/Services/UserPlaylistsService.php [hunk 1]
```diff
@@ -66,11 +76,32 @@ public function delete($id)
...
+     * @param null $limit
+     * @param int $page
+     * @param null $term
+     * @param string $sort
+     * @return mixed|\Railroad\Railcontent\Support\Collection|null
      */
-    public function getUserPlaylist($userId, $playlistType, $brand = null)
-    {
-        return $this->userPlaylistsRepository->getUserPlaylist($userId, $playlistType, $brand);
+    public function getUserPlaylist(
```
Comment (robwillems) [inline src/Services/UserPlaylistsService.php:85]: **suggestion:**
This returns a collection of playlists so i'd expect the naming to be plural.  GetUserPlaylists.

File: migrations/2022_12_12_080000_add_playlists_data_to_playlists_table.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            config('railcontent.table_prefix') . 'playlist_likes',
+            function(Blueprint $table) {
+                $table->increments('id');
+                $table->integer('playlist_id')->index();
+                $table->integer('user_id')->index();
+                $table->dateTime('created_at')->index();
+            });
+
+        Schema::connection(config('railcontent.database_connection_name'))->create(
+            config('railcontent.table_prefix') . 'pinned_playlists',
```
Comment (robwillems) [inline migrations/2022_12_12_080000_add_playlists_data_to_playlists_table.php:48]: **suggestion:**

Ideally the names would be consistent to make it easy to see which tables are used by the playlist feature.
Suggestion is to rename the last two tables, but maybe there is some context here I don't fully understand.

railcontent_content_playlists
railcontent_user_playlists
railcontent_user_playlist_content
railcontent_user_playlist_likes
railcontent_user_playlist_pinned

File: src/Controllers/MyListJsonController.php [hunk 3]
```diff
@@ -214,4 +209,481 @@ public function getMyLists(Request $request)
...
+     * @return mixed
+     */
+    public function createPlaylist(PlaylistCreateRequest $request)
+    {
+        $playlist = $this->userPlaylistsService->create([
+                                                            'user_id' => auth()->id(),
+                                                            'type' => 'user-playlist',
+                                                            'brand' => $request->get('brand')
+                                                                ??
+                                                                config('railcontent.brand'),
```
Comment (robwillems) [inline src/Controllers/MyListJsonController.php:224]: config('railcontent.brand') is 'musora'.  Is this a case we need to deal with now?  Should we just make brand required in the request?

File: src/Controllers/MyListJsonController.php [hunk 4]
```diff
@@ -214,4 +209,481 @@ public function getMyLists(Request $request)
...
+            $limit,
+            $page,
+            $request->get('term'),
+            $sort
+        );
+
+        $itemId = $request->get('content_id');
+        if ($itemId) {
+            foreach ($playlists as $index => $playlist) {
+                $playlistItem = $this->userPlaylistsService->existsContentIdInPlaylist($playlist['id'], $itemId);
```
Comment (robwillems) [inline src/Controllers/MyListJsonController.php:326]: Would it be worth optimizing this to use a single query?

File: src/Services/UserPlaylistsService.php [hunk 5]
```diff
@@ -134,12 +170,543 @@ public function countUserPlaylistContents($playlistId, $contentType = [])
...
+                    'start_second' => $startSecond,
+                    'end_second' => $endSecond,
+                    'created_at' => Carbon::now()
+                        ->toDateTimeString(),
+                ];
+
+                $this->userPlaylistContentRepository->createOrUpdatePlaylistContentAndReposition(null, $input);
+
+                if ($importAllAssignments) {
+                    $this->addSounsliceAssignments(
```
Comment (robwillems) [inline src/Services/UserPlaylistsService.php:280]: Naming:  addSoundSliceAssignments

**Review by robwillems — COMMENTED** [review]
I had a quick skim through and noted a few minor things.  

I'll take a deeper stepping through the code tomorrow.  This change is very large so it will take some time to go through each endpoint properly.

File: src/Repositories/UserPlaylistContentRepository.php [hunk 6]
```diff
@@ -62,22 +54,32 @@ public function getUserPlaylistContents($playlistId, $contentType = [], $limit =
...
+            $query = $query->orderBy(config('railcontent.table_prefix').'user_playlist_content.position', 'asc');
+        }
+
         $contentRows =
-            $query->orderBy(config('railcontent.table_prefix').'user_playlist_content.id', 'desc')
-                ->get()
+            $query->get()
                 ->toArray();
 
         $extraData = $this->geExtraDataInOldStyle(['data', 'instructor', 'video'], $contentRows);
```
Comment (robwillems) [inline src/Repositories/UserPlaylistContentRepository.php:79]: getExtraDataInOldStyle

Comment (robwillems) [inline src/Repositories/UserPlaylistContentRepository.php:79]: This function could be optimized.  Currently runs a query for each key.

File: src/Repositories/PinnedPlaylistsRepository.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    public function query()
+    {
+        return $this->connection()
+            ->table(config('railcontent.table_prefix').'pinned_playlists');
+    }
+
+    /**
+     * @return array|mixed[]
+     */
+    public function getMyPinnedPlaylists()
```
Comment (robwillems) [inline src/Repositories/PinnedPlaylistsRepository.php:18]: This query is pretty expensive, and it looks like its used for whats shown in the navigation menu, so this should be optimized or cached at some level.

File: src/Repositories/PinnedPlaylistsRepository.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                        config('railcontent.table_prefix').'user_playlists'.'.id'
+                    );
+                })
+                ->leftjoin(config('railcontent.table_prefix').'user_playlist_content', function (JoinClause $join) {
+                    $join->on(
+                        config('railcontent.table_prefix').'pinned_playlists'.'.playlist_id',
+                        '=',
+                        config('railcontent.table_prefix').'user_playlist_content'.'.user_playlist_id'
+                    )
+                        ->where(config('railcontent.table_prefix').'user_playlist_content.position', '=', 1);
```
Comment (robwillems) [inline src/Repositories/PinnedPlaylistsRepository.php:41]: What is this for?  user_playlist_content.position = 1?  
Is this for if a playlist is started or completed?

**Review by robwillems — COMMENTED** [review]
I found a few more things.
A lot of optimization depends on where these will be used so I was using musora-web-platform project-playlists-v2 to figure some of it out.
I also think some optimizations are required in PlaylistDecorator and the nav loading, but that could wait until the musora-web-platform Code review is posted.

File: migrations/2022_12_12_080000_add_playlists_data_to_playlists_table.php (hunk 2 above)
Comment (roxanariza) [reply migrations/2022_12_12_080000_add_playlists_data_to_playlists_table.php:48]: Good observation! I will update the tables names

**Pull request merged**